cartridge/roles.lua

```diff
@@ -17,108 +17,148 @@ local errors = require('errors')
+    checks({
+        roles_by_number = 'table',
+        roles_by_role_name = 'table',
+        roles_by_module_name = 'table',
+    }, 'string')
+
+    if ctx.roles_by_module_name[module_name] then
         -- Already loaded
-        return mod
+        return ctx.roles_by_module_name[module_name]
     end
 
-    local mod, err = RegisterRoleError:pcall(require, module_name)
-    if not mod then
-        return nil, e(err)
-    elseif type(mod) ~= 'table' then
-        return nil, e('Module %q must return a table', module_name)
+    local M, err = RegisterRoleError:pcall(require, module_name)
+    if err ~= nil then
+        return nil, err
+    elseif type(M) ~= 'table' then
+        return nil, RegisterRoleError:new(
+            'Module %q must return a table, got %s', module_name, type(M)
+        )
     end
 
-    if mod.role_name == nil then
-        mod.role_name = module_name
+    local role = {
+        M = M,
+        deps = {},
+        role_name = M.role_name or module_name,
+        module_name = module_name,
+    }
+
+    if type(role.role_name) ~= 'string' then
+        return nil, RegisterRoleError:new(
+            'Module %q role_name must be a string, got %s',
+            module_name, type(role.role_name)
+        )
     end
 
-    if type(mod.role_name) ~= 'string' then
-        return nil, e('Module %q role_name must be a string', module_name)
+    local dependencies = M.dependencies or {}
+    if type(dependencies) ~= 'table' then
+        return nil, RegisterRoleError:new(
+            'Module %q dependencies must be a table, got %s',
+            module_name, type(dependencies)
+        )
     end
 
-    if vars.known_roles[mod.role_name] ~= nil then
-        return nil, e('Role %q name clash', mod.role_name)
+    if ctx.roles_by_role_name[role.role_name] ~= nil then
+        return nil, RegisterRoleError:new(
+            'Role %q name clash between %s and %s',
+            role.role_name, module_name,
+            ctx.roles_by_role_name[role.role_name].module_name
+        )
     end
 
-    local dependencies = mod.dependencies or {}
-    if type(dependencies) ~= 'table' then
-        return nil, e('Module %q dependencies must be a table', module_name)
-    end
+    ctx.roles_by_module_name[module_name] = role
+    ctx.roles_by_role_name[role.role_name] = role
+
+    for _, dep in ipairs(dependencies) do
+        local dep_role, err = register_role(ctx, dep)
+        if not dep_role then
+            return nil, err
+        end
 
-    vars.roles_dependencies[mod.role_name] = {}
-    vars.roles_dependants[mod.role_name] = {}
-    vars.known_roles[mod.role_name] = mod
+        if not utils.table_find(role.deps, dep_role) then
+            table.insert(role.deps, dep_role)
+        end
 
-    local function deps_append(tbl, deps)
-        for _, dep in pairs(deps) do
-            if not utils.table_find(tbl, dep) then
-                table.insert(tbl, dep)
+        -- Inherit subdependencies
```

Comment:
What's about the third-level (and more deep) subdependencies?

Reply:
Seems it should be done with simple recursion function

Reply:
Or is this code really needed?

Reply:
Third-level dependensies are OK because this code copies them from previous level.
It's needed and it's covered with tests: https://github.com/tarantool/cartridge/blob/2bd23ad16a028cf113a1b1ecd2317e0c6eeadfa8/test/unit/roledeps_test.lua#L133-L138